File: frontend/src/api/generate.ts
import { apiClient } from './client'

export interface SignalInfo { name: string; width: number; role: string }
export interface GenerateRequest {
  text: string
  code_type: string
  protocol?: string
  clk?: string
  rst?: string
  rst_polarity?: string
  signals?: SignalInfo[]
}
export interface RAGCandidate { template_id: string; name: string; score: number }
export interface GenerateResponse {
  status: string
  confidence: number
  template_id: string
  template_version: string
  cache_hit: boolean
  intent_cache_hit: boolean
  rag_candidates: RAGCandidate[]
  params_used: Record<string, unknown>
  code: string
}

// ── 方案 3 两步式 schema ────────────────────────────────────────────
export type ParamSource = 'signal_list' | 'regex' | 'llm' | 'default' | 'semantic_fallback' | 'placeholder'
export type ConfidenceSource = 'llm_step1' | 'rag_fallback' | 'intent_cache' | 'keyword_supplement'

export interface ParamWithSource {
  value: string | number | string[]
  source: ParamSource
  required: boolean
  description: string
  type: string
  sanitized?: boolean              // 后端做过 SV 标识符清洗
  expr_type?: string               // 模板 YAML 声明的语法类型（驱动前端 dispatch 校验）
  validation_error?: string        // sv_boolean_expr / sv_bins_expr 校验失败时的错误描述
}

export interface RAGCandidateWithParams {
  template_id: string
  name: string
  score: number
  parameters: Array<Record<string, unknown>>
}

export interface PreviewResponse {
  template_id: string
  template_name: string
  template_version: string
  confidence: number
  confidence_source: ConfidenceSource
  rag_candidates: RAGCandidateWithParams[]
  params: Record<string, ParamWithSource>
  intent_hash: string
  normalized_intent: string
  quick_render: boolean
}

export interface RenderConfirmedRequest {
  template_id: string
  template_version: string
  params: Record<string, unknown>
  intent_hash: string
  confidence: number
  confidence_source: ConfidenceSource
  normalized_intent: string
  original_intent: string
  rag_candidates: Array<{ template_id: string; name: string; score: number }>
  code_type: string
}

export type GenerationMode = 'rag' | 'llm_direct'

export interface RenderConfirmedResponse {
  code: string
  cache_hit: boolean
  generation_record_id?: string | null
  // FEAT-11 Stage 2：默认 'rag'；前端用此决定是否显示 LLM fallback 按钮
  generation_mode?: GenerationMode
}

export interface LLMFallbackResponse {
  code: string
  generation_record_id: string
  generation_mode: 'llm_direct'
  cache_hit: boolean
}

export const generateApi = {
  generate: async (req: GenerateRequest) => {
    // legacy 一步式（保留兼容用）。thinking 模型 worst case 3 次 LLM × 60s。
    const res = await apiClient.post<GenerateResponse>('/generate', req, { timeout: 300000 })
    return res.data
  },
  // 方案 3 两步式
  preview: async (req: GenerateRequest) => {
    // thinking 模型（GLM-4.7 等）单次推理 20-60s，三次累加 60-180s，给 5min 留 buffer。
    const res = await apiClient.post<PreviewResponse>('/generate/preview', req, { timeout: 300000 })
    return res.data
  },
  renderConfirmed: async (req: RenderConfirmedRequest) => {
    // 渲染只调 Jinja2，<1s；用户参数可能略慢但不会超 30s
    const res = await apiClient.post<RenderConfirmedResponse>('/generate/render', req, { timeout: 30000 })
    return res.data
  },
  // legacy 重渲染（仅渲染不写库，不传 intent_hash）
  render: async (template_id: string, template_version: string, params: Record<string, unknown>) => {
    const res = await apiClient.post<{ code: string }>('/generate/render', { template_id, template_version, params })
    return res.data
  },
  codeTypes: async () => {
    const res = await apiClient.get<{ id: string; display_name: string }[]>('/generate/code-types')
    return res.data
  },
  // FEAT-11 Stage 2：用户点"对生成结果不满意？尝试 LLM 直接生成"时触发
  llmFallback: async (generation_record_id: string) => {
    // freeform 生成走单次 LLM 调用：GLM-4.7 thinking off p50≈3s；为兼容偏慢的
    // thinking-on 模型（曾观察到 60-120s 偶发），设 120s timeout。比 preview 端
    // 的 5min 短（preview 是三段 LLM 累加），但对单次调用足够。
    const res = await apiClient.post<LLMFallbackResponse>(
      '/generate/llm-fallback',
      { generation_record_id },
      { timeout: 120000 },
    )
    return res.data
  },
}

File: backend/app/schemas/generate.py
from __future__ import annotations
import re
from typing import Literal
from pydantic import BaseModel, field_validator

_SIGNAL_NAME_RE = re.compile(r'^[A-Za-z_][A-Za-z0-9_]*$')


class SignalInfo(BaseModel):
    name: str
    width: int
    role: str

    @field_validator("name")
    @classmethod
    def _validate_signal_name(cls, v: str) -> str:
        if not _SIGNAL_NAME_RE.match(v):
            raise ValueError(f"信号名 '{v}' 包含非法字符，只允许字母、数字和下划线，且不能以数字开头")
        return v


class GenerateRequest(BaseModel):
    text: str
    code_type: str
    protocol: str | None = None
    clk: str = "clk"
    rst: str = "rst_n"
    rst_polarity: str = "低有效"
    signals: list[SignalInfo] = []


class RAGCandidate(BaseModel):
    template_id: str
    name: str
    score: float


class GenerateResponse(BaseModel):
    status: str
    confidence: float | None = None
    template_id: str | None = None
    template_version: str | None = None
    cache_hit: bool = False
    intent_cache_hit: bool = False
    rag_candidates: list[RAGCandidate] = []
    params_used: dict | None = None
    extracted_params: dict | None = None
    code: str | None = None


class RenderRequest(BaseModel):
    """两步式流程的第二步：用户确认参数后渲染。

    两种调用模式：
    - **方案 3 路径**（前端两步式 UI）：传 intent_hash + confidence + normalized_intent
      + rag_candidates，由 pipeline_render 写 GenerationRecord 与缓存
    - **legacy 重渲染路径**：仅传 template_id + template_version + params，
      不写 GenerationRecord（参考既有 /render 老用法）
    """
    template_id: str
    template_version: str
    params: dict
    # 方案 3 透传字段（前端 preview 阶段拿到后传回，让 render 完成 GenerationRecord 写入）
    intent_hash: str | None = None
    confidence: float = 0.0
    confidence_source: str = ""              # "llm_step1" | "rag_fallback" | "intent_cache"
    normalized_intent: str = ""
    original_intent: str = ""
    rag_candidates: list[dict] = []
    code_type: str = ""                      # 用于 GenerationRecord


class RenderResponse(BaseModel):
    code: str
    cache_hit: bool = False


# ── 方案 3：两步式 preview 端点 schema ──────────────────────────────────────

class ParamWithSource(BaseModel):
    """单个参数的预填充值 + 来源标识，用于前端 5 色徽标显示。"""
    value: str | int | float | list[str]
    source: Literal["signal_list", "regex", "llm", "default", "placeholder"]
    required: bool = True
    description: str = ""
    type: str = "string"                     # 模板里声明的 type（用于前端校验）


class RAGCandidateWithParams(BaseModel):
    """RAG 候选（含 parameters 列表，供前端切换模板时重新渲染表单用）。"""
    template_id: str
    name: str
    score: float
    parameters: list[dict] = []


class PreviewResponse(BaseModel):
    """两步式流程的第一步响应：模板推荐 + 参数预填充。"""
    template_id: str
    template_name: str
    template_version: str
    confidence: float
    confidence_source: Literal["llm_step1", "rag_fallback", "intent_cache"]
    rag_candidates: list[RAGCandidateWithParams]
    params: dict[str, ParamWithSource]
    intent_hash: str
    normalized_intent: str
    quick_render: bool = False               # intent_cache 命中 → True，前端跳过确认面板


class BatchUploadResponse(BaseModel):
    job_id: str
    total_rows: int
    code_type: str
    status: str


class BatchStatusResponse(BaseModel):
    job_id: str
    status: str
    total_rows: int
    completed_rows: int
    progress: float
    result_url: str | None = None
    error_message: str | None = None


class PreflightRowResult(BaseModel):
    row_id: str
    estimated_confidence: float
    top_match: dict | None = None


class PreflightResponse(BaseModel):
    results: list[PreflightRowResult]
